experimental/AlgebraicStatistics/src/IndexedRing.jl

```diff
@@ -0,0 +1,77 @@
+# this should find a home outside of AlgStats at some point
+struct IndexedRing{T, U} <: NCRing
+  R::Ring
+  index_to_gen::Dict{T, U}
+  gen_to_index::Dict{U, T}
+
+  function IndexedRing(S::Ring, varnames::Vector{<:VarName}; kw...)
+    R, x =  polynomial_ring(S, varnames; kw...)
+    index_to_gen = Dict{Int, elem_type(R)}(i => x[i] for i in 1:ngens(R))
+    gen_to_index = Dict{elem_type(R), Int}(v => k for (k, v) in index_to_gen)
+    return new{Int, elem_type(R)}(R, index_to_gen, gen_to_index)
+  end
+
+  function IndexedRing(S::Ring, varnames::Pair{String, <: Array{T}}; kw...) where T
+    gen_names = ["$(varnames.first)[$(join(x, ","))]" for x in varnames.second]
+    R, r =  polynomial_ring(S, gen_names; kw...)
+    U = elem_type(R)
+    index_to_gen = Dict{T, U}(t => r[i] for (i, t) in enumerate(varnames.second))
+    gen_to_index = Dict{U, T}(v => k for (k, v) in index_to_gen)
+    return new{T, U}(R, index_to_gen, gen_to_index)
+  end
+end
+
+_ring(R::IndexedRing) = R.R
+symbols(R::IndexedRing) = symbols(_ring(R))
+base_ring(R::IndexedRing) = base_ring(_ring(R))
+elem_type(R::IndexedRing) = elem_type(_ring(R))
```

Comment:
I think this really should be
```suggestion
base_ring(R::IndexedRing) = R.R
symbols(R::IndexedRing) = symbols(base_ring(R))
coefficient_ring(R::IndexedRing) = coefficient_ring(base_ring(R))
elem_type(R::IndexedRing) = elem_type(base_ring(R))
```
but of course that would be quite breaking. As a middle ground, I'd strongly suggest to at least define
```julia
coefficient_ring(R::IndexedRing) = coefficient_ring(_ring(R))
```
for now and then switch code from using `base_ring` to `coefficient_ring`.